### DISCUSSION TITLE
Access ctx inside sessionMiddleware getSession etc...

### DISCUSSION THREAD

**Discussion by cj** [blitz-js/blitz#2366] [Ideas]
It would be great to be able to access the ctx inside `sessionMiddleware` when managing sessions, that way we can save additional user information!

I have also noticed `createSession` is not triggered for anonymous users.

**Comment by cj** [comment]
Ok so when I call `await session.$setPrivateData({ integration: { ...integration, ...input } })` `createSession` is triggered, but this is all it contains:

```
{ createSession: { handle: '4avxVPgtdWW5TjGp7gN2HTBC2s7wuahx:ajwt' } }
```

Right after it calls updateSession which only includes:

```
{
  updateSession: {
    sessionHandle: '4avxVPgtdWW5TjGp7gN2HTBC2s7wuahx:ajwt',
    session: {
      privateData: 'this string of private data'
    }
  }
}
```

It seems to be missing all the information from:

```
        await global.sessionConfig.createSession({
          expiresAt,
          handle,
          userId: newPublicData.userId,
          hashedSessionToken: hash256(sessionToken),
          antiCSRFToken,
          publicData: JSON.stringify(newPublicData),
          privateData: JSON.stringify(newPrivateData),
        })
```

**Reply by flybayer** [reply]
Anonymous sessions are not saved in the DB by default. They are stored in cookies as a JWT.

If you setPrivateData, then it will be saved in the DB. It's correct that it doesn't have all the other things because the other things are stored inside the JWT.

**Reply by cj** [reply]
That makes sense.  Maybe we should change the behaviour of `setPrivateData` to make it send just one `createSession` event with the privateData, rather that one `createSession` with only the handle and then an 'updateSession' with the pivateData?

**Reply by flybayer** [reply]
Yeah possibly we can optimize that!

**Comment by flybayer** [comment]
Can you give an example of how you want to access ctx there?

**Reply by cj** [reply]
So when I am creating the session I can store the users browser and ip information.

**Reply by flybayer** [reply]
Ok yeah we can add that. You can open an issue
